### DISCUSSION TITLE
search_for in Fitz

### DISCUSSION THREAD

**Discussion by maivan-hoa** [pymupdf/PyMuPDF#3251] [Looking for help]
I have two pdf files: `File_A.pdf` and `File_B.pdf`. These two files have the same content, I just reversed the sentence order in the two files. But the `search_for` function doesn't work with `File_A`.

Below is the code that I use:
```python
import fitz
import os

input_file = r'./path_to_file.pdf'
list_text = ['Quy định tại Điều 7, Điều 8 và Điều 9 Thông tư này.', 'Quy định tại Điều 4, Điều 5 và Điều 6 Thông tư này.']

doc = fitz.open(input_file)

for page in doc:
    for text in list_text:
        ### SEARCH
        rlist = page.search_for(text)
        print('TEXT: ', text)
        print('RLIST: ', rlist)
```

- When using the code with `File_A.pdf`, I get the following result:
```
TEXT:  Quy định tại Điều 7, Điều 8 và Điều 9 Thông tư này.
RLIST:  []

TEXT:  Quy định tại Điều 4, Điều 5 và Điều 6 Thông tư này.
RLIST:  []
```

- When using the code with `File_B.pdf`, I get the following result:
```
TEXT:  Quy định tại Điều 7, Điều 8 và Điều 9 Thông tư này.
RLIST:  [Rect(85.30000305175781, 78.54861450195312, 95.26360321044922, 96.95781707763672), Rect(95.5, 80.78265380859375, 124.55408477783203, 96.06507873535156), Rect(124.31999969482422, 78.54861450195312, 146.1239776611328, 96.95781707763672), Rect(145.9499969482422, 80.78265380859375, 379.6589660644531, 96.06507873535156), Rect(380.20001220703125, 78.54861450195312, 383.6500244140625, 96.95781707763672)]

TEXT:  Quy định tại Điều 4, Điều 5 và Điều 6 Thông tư này.
RLIST:  [Rect(85.30000305175781, 58.11856460571289, 113.34999084472656, 76.52776336669922), Rect(113.5199966430664, 60.35260009765625, 124.55408477783203, 75.63502502441406), Rect(124.31999969482422, 58.11856460571289, 146.1239776611328, 76.52776336669922), Rect(145.9499969482422, 60.35260009765625, 152.07508850097656, 75.63502502441406), Rect(151.9499969482422, 58.11856460571289, 159.00039672851562, 76.52776336669922), Rect(159.17999267578125, 60.35260009765625, 179.10508728027344, 75.63502502441406), Rect(178.97000122070312, 58.11856460571289, 203.97557067871094, 76.52776336669922), Rect(204.1999969482422, 60.35260009765625, 214.16590881347656, 75.63502502441406), Rect(214.42999267578125, 58.11856460571289, 218.26638793945312, 76.52776336669922), Rect(218.02999877929688, 60.35260009765625, 224.15509033203125, 75.63502502441406), Rect(224.02000427246094, 58.11856460571289, 262.30023193359375, 76.52776336669922), Rect(262.4800109863281, 60.35260009765625, 282.39508056640625, 75.63502502441406), Rect(282.29998779296875, 58.11856460571289, 348.5863952636719, 76.52776336669922), Rect(348.3500061035156, 60.35260009765625, 366.6212158203125, 75.63502502441406), Rect(366.9800109863281, 58.11856460571289, 383.622802734375, 76.52776336669922)]
```

I also attached these two files below. Look forward to the help of everyone!
[File_A.pdf](https://github.com/pymupdf/PyMuPDF/files/14553348/File_A.pdf)
[File_B.pdf](https://github.com/pymupdf/PyMuPDF/files/14553349/File_B.pdf)

**Comment by JorjMcKie** [comment]
Interesting problem!
The answer is that you did **_not_** just change the sequence of the sentences. No idea what you did, but look here. I extracted the text from both pages splitting it into lines. Then compared the lines that **_look_** equal:
```python
print(f"{texta[1]=}")  # line 2 from File_A
texta[1]='Quy định tại Điều 4, Điều 5 và Điều 6 Thông tư này. '
print(f"{textb[0]=}")  # line 1 from File_B
textb[0]='Quy định tại Điều 4, Điều 5 và Điều 6 Thông tư này. '
texta[1] == textb[0]  # !!!
False
len(texta[1])  # even different lengths!
57
len(textb[0])
52
```

**Reply by maivan-hoa** [reply]
Thank you for your feedback.
I simply use copy and paste in the Microsoft Word editor. Below is the docx file I used for editing
[File_A.pdf](https://github.com/pymupdf/PyMuPDF/files/14558060/File_A.pdf)
[File_B.pdf](https://github.com/pymupdf/PyMuPDF/files/14558061/File_B.pdf)
[File_test.docx](https://github.com/pymupdf/PyMuPDF/files/14558062/File_test.docx)

**Reply by JorjMcKie** [reply]
I cannot check what Word is doing in this case.
Admittedly, Vietnamese text is full of accents and other character modifiers. I imagine, that copying text previously written does not mean that the clipboard content is exactly the same - down to the bit level.
So when pasting the clipboard again, it may **_look_** identical but not **_be_** identical at the low level.

Maybe things are different if you choose a different font - thus taking away some liberty from how Word is doing things ...

I tried the whole thing with LibreOffice:
1. converted file_test.docx to the odt format
2. created 2 different versions by change the sequence via copy-paste.
3. Used LibreOffice to make the two PDFs

This time it did work! Also the created PDFs are much smaller than what Word produced.
[file-a.odt](https://github.com/pymupdf/PyMuPDF/files/14558421/file-a.odt)
[file-a.pdf](https://github.com/pymupdf/PyMuPDF/files/14558422/file-a.pdf)
[file-b.odt](https://github.com/pymupdf/PyMuPDF/files/14558423/file-b.odt)
[file-b.pdf](https://github.com/pymupdf/PyMuPDF/files/14558424/file-b.pdf)

**Reply by JorjMcKie** [reply]
```python
import fitz
doc=fitz.open("file-a.pdf")
print(doc[0].get_text())
Quy định tại Điều 7, Điều 8 và Điều 9 Thông tư này.
Quy định tại Điều 4, Điều 5 và Điều 6 Thông tư này.

text="Quy định tại Điều 4, Điều 5 và Điều 6 Thông tư này."
page=doc[0]
page.search_for(text)
[Rect(85.1500015258789, 83.75079345703125, 375.63818359375, 97.75079345703125)]
doc=fitz.open("file-b.pdf")
page=doc[0]
page.search_for(text)
[Rect(85.1500015258789, 62.80078125, 375.4181823730469, 76.80078125)]
```

**Reply by maivan-hoa** [reply]
I will try, thank you very much

**Reply by maivan-hoa** [reply]
Did you change the font when converting from `docx` to `odt`? I did the same steps as you: convert `docx` file to `odt`, convert `odt` file to `pdf`, but the `search_for` function still doesn't work.
I use Libreoffice command line to do the conversion:
- `[libreoffice_exec(), '--headless', '--convert-to', 'odt', '--outdir', folder_save, path_file_source]`
- `[libreoffice_exec(), '--headless', '--convert-to', 'pdf', '--outdir', folder_save, path_file_source]`

**Reply by JorjMcKie** [reply]
I opened the docx with LibreOffice editor, then saved it in ODT format. Then swapped those two lines and saved it in a different ODT.
Then opened each ODT and executed "Export" directly to PDF.

**Reply by maivan-hoa** [reply]
I just downloaded two files `file-a.pdf` and `file-b.pdf` and tested your code, but the `search_for` function only runs with `file-b.pdf`

```python
import fitz
doc=fitz.open(r'C:\Users\User\Desktop\file-a.pdf')
print(doc[0].get_text())
# Quy định tại Điều 7, Điều 8 và Điều 9 Thông tư này.
# Quy định tại Điều 4, Điều 5 và Điều 6 Thông tư này.

text="Quy định tại Điều 4, Điều 5 và Điều 6 Thông tư này."
page=doc[0]
print(page.search_for(text))
# []
```

I use PyMuPDF 1.23.5

**Reply by maivan-hoa** [reply]
I don't understand why the `search_for` function doesn't always produce results. I use your file `file-a.pdf` and the code below, but when running that code many times, there will be times when `search_for` will return an empty array.
```python
doc=fitz.open(r'./file-a.pdf')
text = doc[0].get_text()
print(text)

print(doc[0].search_for(text))
```

[file-a.pdf](https://github.com/pymupdf/PyMuPDF/files/14583766/file-a.pdf)

**Reply by JorjMcKie** [reply]
No idea. Did the following without problem.
```python
import fitz
doc=fitz.open("file-a.pdf")
page=doc[0]
text=page.get_text().splitlines()[1]
doc.close()
def find_text():
    doc=fitz.open("file-a.pdf")
    page=doc[0]
    r=page.search_for(text)
    if not r:
        print("text not found!")
    doc.close()

    
find_text()
for i in range(100):
    find_text()

    
for i in range(1000):
    find_text()

    
```
